### PR TITLE
CDAP-6372 add spark streaming app

### PR DESCRIPTION
**Pull request by albertshau**
Added apis for new StreamingSource and Windower plugin
types for the data streams application.

Also doing some more refactoring on the spark program so that both
the data streams and data pipeline apps can share almost all the
same logic. Introduced the SparkCollection and SparkPairCollection
interfaces which the main driver performs operations on.
The data pipeline app implements them with RDDs while the data
streams app implements it with DStreams.

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsApp.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  @Override
+  public void configure() {
+    DataStreamsConfig config = getConfig();
+    setDescription("Data Streams Application");
+
+    PipelineSpecGenerator<DataStreamsConfig, DataStreamsPipelineSpec> specGenerator =
+      new DataStreamsPipelineSpecGenerator(
+        getConfigurer(),
+        ImmutableSet.of(StreamingSource.PLUGIN_TYPE),
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsApp.java:39]: Are there no transformations on DataStreams?

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsPipelineSpecGenerator.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  @Override
+  public DataStreamsPipelineSpec generateSpec(DataStreamsConfig config) {
+    long batchIntervalMillis;
+    try {
+      batchIntervalMillis = TimeParser.parseDuration(config.getBatchInterval());
+    } catch (Exception e) {
+      throw new IllegalArgumentException(
+        String.format("Unable to parse batchInterval '%s'", config.getBatchInterval()), e);
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsPipelineSpecGenerator.java]: Let's make sure we won't return the exception in the REST response incase of config issues. The message should be sufficient.

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    properties.put("cask.is.unit.test", String.valueOf(isUnitTest));
+    properties.put("cask.num.sources", String.valueOf(numSources));
+    setProperties(properties);
+  }
+
+  @Override
+  public void beforeSubmit(SparkClientContext context) throws Exception {
+    SparkConf sparkConf = new SparkConf();
+    sparkConf.set("spark.driver.extraJavaOptions", "-XX:MaxPermSize=256m");
+    sparkConf.set("spark.executor.extraJavaOptions", "-XX:MaxPermSize=256m");
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java]: We should make these two configurations configurable.

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsApp.java (hunk 1 above)
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsApp.java:39]: Ignore this comment. I see that we have the other types.

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/test/java/co/cask/cdap/datastreams/DataStreamsTest.java [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+      },
+      1,
+      TimeUnit.MINUTES);
+
+    sparkManager.stop();
+    sparkManager.waitForStatus(false, 10, 1);
+  }
+
+  @Test
+  public void testWindower() throws Exception {
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/test/java/co/cask/cdap/datastreams/DataStreamsTest.java:244]: Is Windowed join supported? We should add a testcase if it is.

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/batch/RDDCollection.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  public <K, V> SparkPairCollection<K, V> flatMapToPair(PairFlatMapFunction<T, K, V> function) {
+    return new PairRDDCollection<>(sec, jsc, datasetContext, sinkFactory, rdd.flatMapToPair(function));
+  }
+
+  @Override
+  public <U> SparkCollection<U> compute(String stageName, SparkCompute<T, U> compute) throws Exception {
+    SparkExecutionPluginContext sparkPluginContext =
+      new BasicSparkExecutionPluginContext(sec, jsc, datasetContext, stageName);
+
+    // TODO: figure out how to do this in a better way...
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/batch/RDDCollection.java]: Can we add JIRAs for these.

**Comment by sreevatsanraman** [comment]
Have minor comments. Fix and merge. LGTM

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/test/java/co/cask/cdap/datastreams/DataStreamsTest.java (hunk 4 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/test/java/co/cask/cdap/datastreams/DataStreamsTest.java:244]: There is no windowed join capability in a single stage. To do it, somebody would first window then join, in which case the two plugin types can be tested separately.

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT4509

**Pull request merged**